### ISSUE TITLE
jbang export --slimjar / --standalone 

### ISSUE THREAD

**Issue by maxandersen** [jbangdev/jbang#619]
currently jbang assumes the app built runs on same machine as its built on.

Could be nice to have export support a "standalone" option (i.e. `jbang export --standalone xyz.jar` where it would not only output a jar, but add a `libs/*` folder with all its dependencies and a matching `xyz`/`xyz.cmd` with the proper argument setup/generated.

besides normally "portability" usecase, it has a usecase in lego mindstorms (ev3) where the host is too slow to build but good enough for running: https://twitter.com/juanantoniobm/status/1341895888034263040

**Comment by quintesse** [comment]
I wonder, does the Class-Path in the MANIFEST take relative paths? That way we might do without the script/batch file.

**Comment by quintesse** [comment]
And another option would be `--fatjar` of course

**Comment by maxandersen** [comment]
It does take relative paths. 

And if by fatjar you mean a shaded/combined jar then sure but that would be a second choice ...

**Comment by quintesse** [comment]
> but that would be a second choice

Just to make sure: I meant that we could have both options.

**Comment by maxandersen** [comment]
Yes. Could have a "style". The slimjar is good for when just syncing update logic and deps can stay tbe same. 

Fat jar is for easy distro.

**Comment by maxandersen** [comment]
Btw. These "exports" are probably the most perfect thing to make external scripts as all the relevant parts would or at least should be available as long as you have the current exported jar. 

They are basically transforms.

**Comment by jabrena** [comment]
Hi, I share few ideas:

https://github.com/ev3dev-lang-java/template-project-gradle/blob/master/config.gradle

https://github.com/ev3dev-lang-java/template-project-gradle/blob/master/build.gradle

https://github.com/ev3dev-lang-java/template-project-gradle/blob/master/gradle/deploy.gradle

cheers

**Comment by jabrena** [comment]
I recommend to put focus on a fat jar solution. It takes few seconds to deploy on EV3 and it is a solution that fit nice for a teenager.

I recommend to allow execute Linux command in front of jar (brickrun time) and set JVM flags. In some scenarios, the jar requires sudo privilegies.

**Comment by maxandersen** [comment]
Wouldn't slim jar be just as fast ?

I'm still waiting on my e3v to arrive but I assumed copy a blob would be quite slower than just the changed files ... assuming actually using dependencies.

**Comment by jabrena** [comment]
Nope, the users doesn’t have that knowledge. The fat jar simplify the operation.

**Comment by jabrena** [comment]
The benefit from jbang is to avoid any IDE.

Donwload from internet one example.
Install jbang
Run example

It is a good scenario, because it is straight forward

**Comment by jabrena** [comment]
—target could be interesting. You could code in Java 15 and generate an artifact for Java 11

**Comment by maxandersen** [comment]
> —target could be interesting. You could code in Java 15 and generate an artifact for Java 11

you have a few options there:

1) use `jbang --java 11` and it will force to use java 11 to build with (it will download java 11 if you don't have it) 

2) use `//JAVAC_OPTION` to set whatever compiler flags you want, include source and target levels.

**Comment by maxandersen** [comment]
> It is a good scenario, because it is straight forward. In 5 minutes, you have your Java program running with 0 knowledge of Java.

I hear what you are saying but none of that changes wether it generates a slim vs fat jar does it ? except the fat jar will be slower to create ? 
 
> it is fair to have the slim option but it will be for very very advanced users, adults basically.

thats what I'm trying to understand ... why would you use a noob care in this case ? 

I guess if you want to do it manually and have to realize you have to copy both the slim.jar and the `lib` folder then its more manual work.

Is that what you have in mind ?

**Comment by jabrena** [comment]
> I guess if you want to do it manually and have to realize you have to copy both the slim.jar and the lib folder then its more manual work.

Exactly. It is a level of complexity that a teenager will scape and the STEAM teacher too.

**Comment by quintesse** [comment]
> I guess if you want to do it manually and have to realize you have to copy both the slim.jar and the lib folder then its more manual work.

That's what I would think too. It's more difficult in a bunch of situations IMO: first is the fact that copying this around means needing more knowledge of *nix command parameters and secondly if you want to share the result with someone it has suddenly become a _lot_ more difficult. You probably need to know how to zip or tar because sending a bunch of files is always problematic.

> except the fat jar will be slower to create ?

is it really going to be that much slower to create? in a way noticeable by the user? Isn't it simply unpacking all deps into a single JAR?

**Comment by jabrena** [comment]
> is it really going to be that much slower to create? in a way noticeable by the user? Isn't it simply unpacking all deps into a single JAR?

No, definitively no. In a modern computer, takes one seconds to create a fatjar.

Using Idea + Gradle + the process to create a fatjar and deploy in the brick, takes 10 seconds with a OSX with ICore5.

**Comment by maxandersen** [comment]
the thing to remember is that a fat jar can't be just lumping all .jar content from multiple jar files into one - things like use of `META-INF/services` would need to be merged somewhow.

See https://product.hubspot.com/blog/the-fault-in-our-jars-why-we-stopped-building-fat-jars for details.

Basically to do this "right" we need something like a shading jar implementation.

**Comment by jabrena** [comment]
In Gradle, I am using this plugin:

``` groovy
plugins {
    id "com.github.johnrengelman.shadow" version "4.0.3"
}
```

**Comment by quintesse** [comment]
Nobody turned this into a Maven library yet? :-)

**Comment by quintesse** [comment]
@jabrena thing is we can't use Maven or Gradle plugins, Jbang doesn't use any of those for compiling, just plain Java. So we'd need something that we can pull in as a Maven dependency (or werite it ourselves if that doesn't exist).

**Comment by jabrena** [comment]
For Maven, I used the following plugin:

```
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-assembly-plugin</artifactId>
                <version>3.1.1</version>

                <configuration>
                    <descriptorRefs>
                        <descriptorRef>jar-with-dependencies</descriptorRef>
                    </descriptorRefs>

                    <archive>
                        <manifestFile>src/main/resources/META-INF/MANIFEST.MF</manifestFile>
                    </archive>
                </configuration>

                <executions>
                    <execution>
                        <id>make-assembly</id>
                        <phase>package</phase>
                        <goals>
                            <goal>single</goal>
                        </goals>
                    </execution>
                </executions>

            </plugin>
```

Reference:
https://github.com/broker-game/template-project-maven/blob/master/pom.xml

**Comment by jabrena** [comment]
Hi @quintesse,

Got it . 

Maybe `John Rengelman` knows the answer. 
https://github.com/johnrengelman/shadow

I will ask him today.

**Comment by maxandersen** [comment]
yes, its clear we can make all of it work - just that fat/shaded jar requires (much) more work than the slim jar appoarch to actually work reliably.

I'll start with slim jar as its straightforward and will work in all cases; definitely open for PR's if we find a lib or someone implements fat/shading jar properly in standalone java.

**Comment by jabrena** [comment]
If the `slim jar` mode:

- remove folder for libs
- create a folder for libs
- move dependencies
- deploy main jar
- execute

In that way, any deployment is atomic so, it is similar to a fatjar but in another way. :)
